From 7bf9e939f0535f1a727182c902baa9f6118208e8 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Wed, 18 Jan 2023 11:47:10 +0100
Subject: [PATCH 275/841] feat(stm32mp2): handle DDR power supplies

Modify platform driver to handle the DDR power supplies when
a PMIC is present.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git c44e45ad7437a381c6fb5aaf990e22f4715f92c4 feat(stm32mp2): handle DDR power supplies]
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Change-Id: I98df132a63c2ad351d4dae949f5dbb831cc40637
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/285864
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 plat/st/stm32mp2/plat_ddr.c | 107 ++++++++++++++++++++++++++++++++++++
 1 file changed, 107 insertions(+)
 create mode 100644 plat/st/stm32mp2/plat_ddr.c

diff --git a/plat/st/stm32mp2/plat_ddr.c b/plat/st/stm32mp2/plat_ddr.c
new file mode 100644
index 000000000..47472101c
--- /dev/null
+++ b/plat/st/stm32mp2/plat_ddr.c
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2023, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+
+#include <drivers/delay_timer.h>
+#include <drivers/st/stm32mp_ddr.h>
+#include <drivers/st/stm32mp_pmic2.h>
+
+#include <platform_def.h>
+
+static int pmic_ddr_power_init(enum ddr_type ddr_type)
+{
+#if STM32MP_DDR4_TYPE
+	int status;
+	struct rdev *vdd_ddr, *vref_ddr, *vtt_ddr, *vpp_ddr;
+
+	/*
+	 * DDR power on sequence is:
+	 * enable VPP_DDR
+	 * wait 2ms
+	 * enable VREF_DDR, VTT_DDR, VPP_DDR
+	 */
+
+	/*
+	 * FIXME: to be replaced by regulator_get_by_supply_name(... "vpp")
+	 * vpp-supply = <&vpp_ddr>; in the device-tree
+	 */
+	/* vpp_ddr */
+	vpp_ddr = regulator_get_by_name("ldo5");
+	if (vpp_ddr == NULL) {
+		return -ENOENT;
+	}
+	status = regulator_set_min_voltage(vpp_ddr);
+	if (status != 0) {
+		return status;
+	}
+
+	/* vdd_ddr */
+	vdd_ddr = regulator_get_by_name("buck6");
+	if (vdd_ddr == NULL) {
+		return -ENOENT;
+	}
+	status = regulator_set_min_voltage(vdd_ddr);
+	if (status != 0) {
+		return status;
+	}
+
+	/* vref_ddr */
+	vref_ddr = regulator_get_by_name("refddr");
+	if (vref_ddr == NULL) {
+		return -ENOENT;
+	}
+
+	/* vtt_ddr */
+	vtt_ddr = regulator_get_by_name("ldo3");
+	if (vtt_ddr == NULL) {
+		return -ENOENT;
+	}
+	status = regulator_set_flag(vtt_ddr, REGUL_SINK_SOURCE);
+	if (status != 0) {
+		return status;
+	}
+
+	status = regulator_enable(vpp_ddr);
+	if (status != 0) {
+		return status;
+	}
+
+	/* could be set via enable_ramp_delay on vpp_ddr */
+	udelay(2000);
+
+	status = regulator_enable(vdd_ddr);
+	if (status != 0) {
+		return status;
+	}
+
+	status = regulator_enable(vref_ddr);
+	if (status != 0) {
+		return status;
+	}
+
+	status = regulator_enable(vtt_ddr);
+	if (status != 0) {
+		return status;
+	}
+#else
+	ERROR("DDR type no supported with PMIC\n");
+	panic();
+#endif
+
+	return 0;
+}
+
+int stm32mp_board_ddr_power_init(enum ddr_type ddr_type)
+{
+	if (dt_pmic_status() > 0) {
+		return pmic_ddr_power_init(ddr_type);
+	}
+
+	return 0;
+}
-- 
2.47.2

